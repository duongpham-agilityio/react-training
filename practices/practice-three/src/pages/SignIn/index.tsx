import {
  Box,
  Button,
  Center,
  Heading,
  Square,
  Text,
  VStack,
} from '@chakra-ui/react';
import { memo, useCallback, useMemo } from 'react';
import { useNavigate } from 'react-router-dom';

// HOCs
import { withIsUnAuth } from '@/hocs';

// Hooks
import { useForm, useHandleAuth } from '@/hooks';

// Stores
import { TAuthStore, useAuthStore } from '@/stores';

// Constants
import { ROUTES } from '@/constants';

// Components
import { InputForm } from '@/components/common';

export interface IFormData {
  email: string;
  password: string;
}

const Component = () => {
  const { formData, onChange } = useForm<IFormData>({
    email: '',
    password: '',
  });
  const { error, onLogin } = useHandleAuth<IFormData>();
  const setAuth = useAuthStore((state: TAuthStore) => state.setIsAuth);
  const redirect = useNavigate();

  const handleChangeValue = useCallback(
    (value: string, name?: string) => {
      const key = (name ?? 'name') as keyof IFormData;

      onChange(value, key);
    },
    [onChange],
  );

  const handleVerifySuccessFull = useCallback(() => {
    setAuth();
    redirect(ROUTES.ROOT, { replace: true });
  }, [redirect, setAuth]);

  const handleLogin = useCallback(async () => {
    onLogin(formData, handleVerifySuccessFull);
  }, [formData, handleVerifySuccessFull, onLogin]);

  // !ISSUES: Because it is currently being re-render, I must use useMemo to remember
  const renderFormHeader = useMemo(
    () => (
      <>
        {/* Heading */}
        <Center h="fit-content" gap={2} py={6}>
          <Box w="1.5" h={39} bg="yellow.10" />
          <Heading fontSize={32}>CRUD OPERATIONS</Heading>
        </Center>
        {/* Title */}
        <Box textAlign="center" py={4}>
          <Heading fontSize={22} py={1}>
            Sign In
          </Heading>
          <Text color="gray.20" fontSize={14} fontWeight="regular">
            Enter your credentials to access your account
          </Text>
        </Box>
      </>
    ),
    [],
  );

  return (
    <Square bg="yellow.20" minH="100vh">
      <Box
        w={475}
        h={550}
        borderRadius={20}
        p={18}
        boxShadow="base"
        bg="white"
        alignItems="unset"
      >
        {renderFormHeader}

        {/* Input Form  */}
        <VStack py={10} gap={5}>
          <InputForm
            isError={!!error['email']}
            errorMessage={error['email']}
            value={formData.email}
            label="Email"
            name="email"
            placeholder="Please enter your email"
            onChange={handleChangeValue}
          />
          <InputForm
            isError={!!error['password']}
            errorMessage={error['password']}
            value={formData.password}
            type="password"
            name="password"
            label="Password"
            placeholder="Please enter your password"
            onChange={handleChangeValue}
          />
        </VStack>

        {/* Submit */}
        <Button
          w="full"
          bg="yellow.20"
          textTransform="uppercase"
          fontWeight="regular"
          fontSize={14}
          _hover={{
            boxShadow: 'base',
          }}
          onClick={handleLogin}
        >
          Sign in
        </Button>
      </Box>
    </Square>
  );
};

const SignIn = memo(withIsUnAuth(Component));

export default SignIn;
