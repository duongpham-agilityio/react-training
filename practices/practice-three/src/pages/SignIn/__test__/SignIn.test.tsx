import { fireEvent, render } from '@testing-library/react';

// Pages
import SignIn from '..';

const setup = () => render(<SignIn />);

describe('SignIn', () => {
  it('Match to snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Change input value', () => {
    const { getByPlaceholderText } = setup();
    const inputEmail: HTMLInputElement = getByPlaceholderText(
      'Please enter your email',
    ) as HTMLInputElement;
    const mockValue = 'duong.pham@gmail.com';

    fireEvent.change(inputEmail, {
      target: {
        value: mockValue,
      },
    });

    expect(inputEmail.value).toBe(mockValue);
  });
});
