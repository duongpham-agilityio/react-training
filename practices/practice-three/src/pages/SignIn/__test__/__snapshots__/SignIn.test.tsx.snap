// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Match to snapshot 1`] = `
<div>
  <div
    class="css-1ydyyj0"
  >
    <div
      class="css-wclwv"
    >
      <div
        class="css-c5yaum"
      >
        <div
          class="css-1icsamv"
        />
        <h2
          class="chakra-heading css-idm3bz"
        >
          CRUD OPERATIONS
        </h2>
      </div>
      <div
        class="css-2gduyd"
      >
        <h2
          class="chakra-heading css-cm18fo"
        >
          Sign In
        </h2>
        <p
          class="chakra-text css-pyoze2"
        >
          Enter your credentials to access your account
        </p>
      </div>
      <div
        class="chakra-stack css-14wdlxi"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r0:"
            id="field-:r0:-label"
          >
            Email
          </label>
          <input
            class="chakra-input css-0"
            id="field-:r0:"
            name="email"
            placeholder="Please enter your email"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r1:"
            id="field-:r1:-label"
          >
            Password
          </label>
          <input
            class="chakra-input css-0"
            id="field-:r1:"
            name="password"
            placeholder="Please enter your password"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="chakra-button css-1yxhbnk"
        type="button"
      >
        Sign in
      </button>
    </div>
  </div>
</div>
`;
